### PR TITLE
[Core] cache `is_async_func()` and `sync_to_async()`

### LINKED ISSUE #45848: [Core] Improve performance of actor task execution by amortizing cost of checking for `async` methods
### Description

While investigating https://github.com/ray-project/ray/issues/45777 I noticed an oddity in a flamegraph from the Serve Controller (attached): about 5% of time was being spent in `inspect.getmembers`, with no caller identified in the flamegraph (small-ish block on the right side of the graph).

controller-under-load

I think this is coming from this chunk of code: https://github.com/ray-project/ray/blob/32cddaecac648fe663c2fef2e7ea8f24e03b234c/python/ray/_raylet.pyx#L1800-L1832

And in fact, if you look at the flamegraph, you'll see a bunch of other calls to similar functions from the `inspect` module which I believe map to other calls in this part of the Raylet. I suspect that these calls appear to come from nowhere in the flamegraph because I'm not running with `py-spy --native`, and the caller is a Cython function, so it won't appear in the stack trace.

I doubt these are very relevant in terms of latency, but they show up as a significant fraction of total time here (probably 5-10%) because the Controller is being hammered with RPCs for what is ultimately a cheap function call (see https://github.com/ray-project/ray/issues/45777).

Looking at the actual code, one solution that jumps out to me is caching. For example, https://github.com/ray-project/ray/blob/035224b52558047d2bb21df415ca75b2f91f0d4c/python/ray/_private/async_compat.py#L30-L32 can be cached - a function's `async`-ness is determined at creation time and can't change later under normal circumstances.

At a higher level, though it's possible for https://github.com/ray-project/ray/blob/32cddaecac648fe663c2fef2e7ea8f24e03b234c/python/ray/_raylet.pyx#L1801-L1803 to change after a class is defined (e.g., by dynamically adding a method to it), it seems like a reasonable restriction to base it on whether the class has any statically-defined `async` methods, so this could be calculated once during actor startup and saved for later.

https://github.com/ray-project/ray/blob/035224b52558047d2bb21df415ca75b2f91f0d4c/python/ray/_private/async_compat.py#L35-L44 Could also potentially be cached, so that you don't need to create a fresh wrapper function on every call to the same method.

I won't get too prescriptive and you may have better ways to profile what's happening here than I do so I'll stop there, but TLDR: doing all this checking over and over is surprisingly time-intensive!

### Use case

This would have a marginal latency improvement for any RPCs and seems especially relevant for high-throughput use cases where this becomes a non-trivial fraction of total time spent on a busy actor serving cheap requests (here it's `ServeController.record_handle_metrics`).

### PR DESCRIPTION
**Pull request by JoshKarpel**
## Why are these changes needed?

See https://github.com/ray-project/ray/issues/45848 for details. This should address most of the problems raised in that issue by caching some of the checks - it would still be better to not do the checks so often, but I don't have the knowledge to do that right now :)

## Related issue number

Closes #45848

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tests/test_async.py [hunk 1]
```diff
@@ -34,25 +33,6 @@ def f(n):
     return [f.remote(i) for i in range(5)]
 
 
-def test_is_async_func():
```
Comment (JoshKarpel) [inline python/ray/tests/test_async.py:37]: Moved to a dedicated file for async compat tests

**Comment by JoshKarpel** [comment]
cc @jjyao for review

File: python/ray/_private/async_compat.py [hunk 2]
```diff
@@ -27,13 +28,20 @@ def try_install_uvloop():
...
-def sync_to_async(func):
-    """Convert a blocking function to async function"""
+@lru_cache(maxsize=2**10)
+def has_async_methods(cls: object) -> bool:
+    """Return True if the class has any async methods."""
+    return len(inspect.getmembers(cls, predicate=is_async_func)) > 0
+
+
+@lru_cache(maxsize=2**10)
+def sync_to_async(func) -> bool:
```
Comment (jjyao) [inline python/ray/_private/async_compat.py:43]: The return type is not bool

**Review by jjyao — COMMENTED** [review]
Looks great!

File: python/ray/_private/async_compat.py (hunk 2 above)
Comment (JoshKarpel) [reply python/ray/_private/async_compat.py:43]: D'oh, good catch, I'll fix this tomorrow.

**Review by jjyao — APPROVED** [review]

**Comment by JoshKarpel** [comment]
Thanks!

**Pull request merged**